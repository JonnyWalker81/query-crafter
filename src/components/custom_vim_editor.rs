use async_trait::async_trait;
use color_eyre::eyre::Result;
use crossterm::event::{KeyCode, KeyEvent, KeyModifiers};
use ratatui::{
  backend::Backend,
  layout::Rect,
  style::{Modifier, Style},
  text::{Line, Span},
  widgets::{Block, Borders, Paragraph},
  Frame,
};
use ropey::{Rope, RopeSlice};

use crate::{
  action::Action,
  editor_common::{Mode, Transition},
  editor_component::EditorComponent,
};

pub struct CustomVimEditor {
  mode: Mode,
  content: Rope,
  cursor_row: usize,
  cursor_col: usize,
  selection: Option<(usize, usize)>, // (start_char_idx, end_char_idx)
}

impl Default for CustomVimEditor {
  fn default() -> Self {
    Self { mode: Mode::Normal, content: Rope::new(), cursor_row: 0, cursor_col: 0, selection: None }
  }
}

impl CustomVimEditor {
  fn move_cursor(&mut self, row_offset: isize, col_offset: isize) {
    let num_lines = self.content.len_lines();
    let new_row_signed = self.cursor_row as isize + row_offset;
    let new_col_signed = self.cursor_col as isize + col_offset;

    let new_row = if new_row_signed < 0 {
      0
    } else if new_row_signed as usize >= num_lines {
      num_lines.saturating_sub(1)
    } else {
      new_row_signed as usize
    };

    let line_len = self.content.line(new_row).len_chars();
    let new_col = if new_col_signed < 0 {
      0
    } else if new_col_signed as usize > line_len {
      line_len
    } else {
      new_col_signed as usize
    };

    self.cursor_row = new_row;
    self.cursor_col = new_col;
  }

  fn insert_char(&mut self, c: char) {
    let char_idx = self.content.line_to_char(self.cursor_row) + self.cursor_col;
    self.content.insert_char(char_idx, c);
    self.move_cursor(0, 1);
  }

  fn insert_newline(&mut self) {
    let char_idx = self.content.line_to_char(self.cursor_row) + self.cursor_col;
    self.content.insert_char(char_idx, '\n');
    self.cursor_row += 1;
    self.cursor_col = 0;
  }

  fn delete_char_backwards(&mut self) {
    if self.cursor_col > 0 {
      let char_idx = self.content.line_to_char(self.cursor_row) + self.cursor_col;
      self.content.remove(char_idx - 1..char_idx);
      self.move_cursor(0, -1);
    } else if self.cursor_row > 0 {
      let char_idx = self.content.line_to_char(self.cursor_row) + self.cursor_col;
      self.content.remove(char_idx - 1..char_idx);
      self.move_cursor(-1, self.content.line(self.cursor_row - 1).len_chars() as isize);
    }
  }

  fn delete_char_forwards(&mut self) {
    let char_idx = self.content.line_to_char(self.cursor_row) + self.cursor_col;
    if char_idx < self.content.len_chars() {
      self.content.remove(char_idx..char_idx + 1);
    }
  }

  fn get_comment_style(line: &str) -> &'static str {
    if line.trim_start().starts_with("//") {
      "//"
    } else if line.trim_start().starts_with("#") {
      "#"
    } else if line.trim_start().starts_with("--") {
      "--"
    } else {
      "//" // Default to C-style comment
    }
  }

  fn toggle_comment(&mut self) {
    let (start_line, end_line) = if let Some((start_idx, end_idx)) = self.selection {
      let start_line = self.content.char_to_line(start_idx);
      let end_line = self.content.char_to_line(end_idx);
      (start_line, end_line)
    } else {
      (self.cursor_row, self.cursor_row)
    };

    for i in start_line..=end_line {
      let line = self.content.line(i);
      let line_str = line.to_string();
      let trimmed_line = line_str.trim_start();
      let leading_whitespace_len = line_str.len() - trimmed_line.len();
      let comment_str = Self::get_comment_style(&line_str);

      if trimmed_line.starts_with(comment_str) {
        // Uncomment
        let comment_start_char_idx = self.content.line_to_char(i) + leading_whitespace_len;
        self.content.remove(comment_start_char_idx..comment_start_char_idx + comment_str.len());
      } else {
        // Comment
        let insert_char_idx = self.content.line_to_char(i) + leading_whitespace_len;
        self.content.insert(insert_char_idx, comment_str);
      }
    }
    self.selection = None;
  }

  fn start_selection(&mut self) {
    let current_char_idx = self.content.line_to_char(self.cursor_row) + self.cursor_col;
    self.selection = Some((current_char_idx, current_char_idx));
  }

  fn update_selection(&mut self) {
    if let Some((start_idx, _)) = self.selection {
      let current_char_idx = self.content.line_to_char(self.cursor_row) + self.cursor_col;
      let (new_start, new_end) =
        if current_char_idx < start_idx { (current_char_idx, start_idx) } else { (start_idx, current_char_idx) };
      self.selection = Some((new_start, new_end));
    }
  }

  fn clear_selection(&mut self) {
    self.selection = None;
  }
}

#[async_trait]
impl EditorComponent for CustomVimEditor {
  fn init(&mut self, _area: Rect) -> Result<()> {
    Ok(())
  }

  fn on_key_event(&mut self, key: KeyEvent) -> Result<Option<Action>> {
    match self.mode {
      Mode::Normal => {
        match key.code {
          KeyCode::Char('h') => self.move_cursor(0, -1),
          KeyCode::Char('j') => self.move_cursor(1, 0),
          KeyCode::Char('k') => self.move_cursor(-1, 0),
          KeyCode::Char('l') => self.move_cursor(0, 1),
          KeyCode::Char('i') => self.mode = Mode::Insert,
          KeyCode::Char('v') => {
            self.start_selection();
            self.mode = Mode::Visual;
          },
          KeyCode::Char('x') => self.delete_char_forwards(),
          KeyCode::Char('q') => return Ok(Some(Action::Quit)),
          KeyCode::Char('c') if key.modifiers.contains(KeyModifiers::CONTROL) => return Ok(Some(Action::Quit)),
          KeyCode::Char('g') => self.mode = Mode::Operator('g'), // For gcc/gc
          _ => {},
        }
      },
      Mode::Insert => {
        match key.code {
          KeyCode::Esc => self.mode = Mode::Normal,
          KeyCode::Backspace => self.delete_char_backwards(),
          KeyCode::Enter => self.insert_newline(),
          KeyCode::Char(c) => self.insert_char(c),
          _ => {},
        }
      },
      Mode::Visual => {
        self.update_selection();
        match key.code {
          KeyCode::Esc => {
            self.clear_selection();
            self.mode = Mode::Normal;
          },
          KeyCode::Char('h') => self.move_cursor(0, -1),
          KeyCode::Char('j') => self.move_cursor(1, 0),
          KeyCode::Char('k') => self.move_cursor(-1, 0),
          KeyCode::Char('l') => self.move_cursor(0, 1),
          KeyCode::Char('c') => {
            self.toggle_comment();
            self.mode = Mode::Normal;
          },
          _ => {},
        }
      },
      Mode::Operator('g') => {
        match key.code {
          KeyCode::Char('c') => {
            self.toggle_comment();
            self.mode = Mode::Normal;
          },
          _ => self.mode = Mode::Normal, // Fallback if not 'gc'
        }
      },
      _ => {},
    }
    Ok(None)
  }

  fn draw(&mut self, f: &mut ratatui::Frame, area: Rect) {
    let mut lines = Vec::new();
    let (selection_start, selection_end) =
      if let Some((start, end)) = self.selection { (Some(start), Some(end)) } else { (None, None) };

    for (i, line) in self.content.lines().enumerate() {
      let mut spans = Vec::new();
      let line_str = line.to_string();
      let line_start_char_idx = self.content.line_to_char(i);

      for (j, c) in line_str.chars().enumerate() {
        let mut style = Style::default();
        let char_absolute_idx = line_start_char_idx + j;

        if let (Some(s_start), Some(s_end)) = (selection_start, selection_end) {
          if char_absolute_idx >= s_start && char_absolute_idx < s_end {
            style = style.add_modifier(Modifier::REVERSED);
          }
        }

        if self.cursor_row == i && self.cursor_col == j {
          style = style.add_modifier(Modifier::REVERSED);
        }
        spans.push(Span::styled(c.to_string(), style));
      }
      lines.push(Line::from(spans));
    }

    let block = Block::default().borders(Borders::ALL).title(format!("Custom Vim Editor - {}", self.mode));
    let paragraph = Paragraph::new(lines).block(block);
    f.render_widget(paragraph, area);
  }

  fn get_text(&self) -> String {
    self.content.to_string()
  }

  fn set_text(&mut self, text: &str) {
    self.content = Rope::from(text);
    self.cursor_row = 0;
    self.cursor_col = 0;
    self.selection = None;
  }

  fn as_any(&self) -> &dyn std::any::Any {
    self
  }

  fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
    self
  }
}
