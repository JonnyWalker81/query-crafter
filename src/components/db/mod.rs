pub mod editor;
pub mod handlers;
pub mod helpers;
pub mod models;
pub mod rendering;
pub mod state;

// Re-export commonly used types
pub use models::{DbColumn, DbTable, QueryHistoryEntry, SelectionMode};

use std::{
    collections::HashMap,
    time::Instant,
};

use color_eyre::eyre::Result;
use nucleo;
use ratatui::prelude::*;
use tokio::sync::mpsc::UnboundedSender;

use self::editor::EditorBackend;
use super::{Component, ComponentKind, Frame};
use crate::{
    action::Action,
    autocomplete::AutocompleteState,
    autocomplete_engine::AutocompleteEngine,
    config::Config,
};

const VISIBLE_COLUMNS: usize = 3;

pub struct Db {
    pub command_tx: Option<UnboundedSender<Action>>,
    pub config: Config,
    
    // Tables and navigation
    pub tables: Vec<DbTable>,
    pub selected_table_index: usize,
    pub is_searching_tables: bool,
    pub table_search_query: String,
    pub table_columns_cache: HashMap<String, Vec<DbColumn>>,
    
    // Query editor
    pub editor_backend: EditorBackend,
    pub query_history: Vec<QueryHistoryEntry>,
    pub selected_history_index: usize,
    pub selected_tab: usize,
    pub last_executed_query: Option<String>,
    
    // Query results
    pub selected_headers: Vec<String>,
    pub query_results: Vec<Vec<String>>,
    pub selected_row_index: usize,
    pub selected_cell_index: usize,
    pub detail_row_index: usize,
    pub horizonal_scroll_offset: usize,
    pub selection_mode: SelectionMode,
    
    // Results search/filter
    pub is_searching_results: bool,
    pub results_search_query: String,
    pub filtered_results: Vec<(usize, Vec<String>)>,
    pub filtered_results_index: usize,
    
    // Component state
    pub selected_component: ComponentKind,
    pub show_help: bool,
    pub error_message: Option<String>,
    pub export_status: Option<(String, Instant)>,
    
    // Autocomplete
    pub autocomplete_state: AutocompleteState,
    pub autocomplete_engine: AutocompleteEngine,
    
    // Table info popup
    pub show_table_columns: bool,
    pub show_table_schema: bool,
    pub selected_table_columns: Vec<DbColumn>,
    pub selected_table_schema: String,
    pub table_info_scroll: usize,
    
    // Query loading state
    pub is_query_running: bool,
    pub query_start_time: Option<Instant>,
    pub last_query_execution_time: Option<u64>, // milliseconds from database
    
    // Preview popup state
    pub preview_scroll_offset: u16,
    pub preview_selected_index: usize,
    
    // Row details popup
    pub show_row_details: bool,
    
    // Results matcher for fuzzy search
    pub results_matcher: nucleo::Matcher,
    
    // EXPLAIN view state
    pub is_explain_view: bool,
    pub is_explain_query: bool,
}

impl Default for Db {
    fn default() -> Self {
        Self::new()
    }
}

impl Db {
    pub fn new() -> Self {
        Self::new_with_config(None)
    }
    
    pub fn new_with_config(config: Option<Config>) -> Self {
        let config = config.unwrap_or_default();
        Self {
            command_tx: None,
            config,
            tables: vec![],
            selected_table_index: 0,
            is_searching_tables: false,
            table_search_query: String::new(),
            table_columns_cache: HashMap::new(),
            editor_backend: EditorBackend::default(),
            query_history: helpers::load_query_history(),
            selected_history_index: 0,
            selected_tab: 0,
            last_executed_query: None,
            selected_headers: vec![],
            query_results: vec![],
            selected_row_index: 0,
            selected_cell_index: 0,
            detail_row_index: 0,
            horizonal_scroll_offset: 0,
            selection_mode: SelectionMode::Table,
            is_searching_results: false,
            results_search_query: String::new(),
            filtered_results: vec![],
            filtered_results_index: 0,
            selected_component: ComponentKind::Home,
            show_help: false,
            error_message: None,
            export_status: None,
            autocomplete_state: AutocompleteState::default(),
            autocomplete_engine: AutocompleteEngine::new_builtin(),
            show_table_columns: false,
            show_table_schema: false,
            selected_table_columns: vec![],
            selected_table_schema: String::new(),
            table_info_scroll: 0,
            is_query_running: false,
            query_start_time: None,
            last_query_execution_time: None,
            preview_scroll_offset: 0,
            preview_selected_index: 0,
            show_row_details: false,
            results_matcher: nucleo::Matcher::new(nucleo::Config::DEFAULT),
            is_explain_view: false,
            is_explain_query: false,
        }
    }
}

impl Component for Db {
    fn register_action_handler(&mut self, tx: UnboundedSender<Action>) -> Result<()> {
        self.command_tx = Some(tx);
        Ok(())
    }

    fn register_config_handler(&mut self, config: Config) -> Result<()> {
        self.register_config_handler(config)
    }

    fn init(&mut self, area: Rect) -> Result<()> {
        self.editor_backend.init(area)?;
        Ok(())
    }

    fn handle_events(&mut self, event: Option<crate::tui::Event>) -> Result<Option<Action>> {
        self.handle_events(event)
    }
    
    fn handle_key_events(&mut self, key: crossterm::event::KeyEvent) -> Result<Option<Action>> {
        self.handle_key_events(key)
    }

    fn update(&mut self, action: Action) -> Result<Option<Action>> {
        self.update(action)
    }

    fn draw(&mut self, f: &mut Frame<'_>, area: Rect) -> Result<()> {
        self.draw(f, area)
    }
}