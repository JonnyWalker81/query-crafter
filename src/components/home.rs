use color_eyre::eyre::Result;
use crossterm::event::KeyEvent;
use ratatui::{prelude::*, widgets::*};
use tokio::sync::mpsc::UnboundedSender;

use super::{Component, Frame};
use crate::{action::Action, config::Config};

#[derive(Default)]
pub struct Home {
  command_tx: Option<UnboundedSender<Action>>,
  config: Config,
}

impl Home {
  pub fn new() -> Self {
    Self::default()
  }
}

impl Component for Home {
  fn register_action_handler(&mut self, tx: UnboundedSender<Action>) -> Result<()> {
    self.command_tx = Some(tx);
    Ok(())
  }

  fn register_config_handler(&mut self, config: Config) -> Result<()> {
    self.config = config;
    Ok(())
  }

  fn update(&mut self, action: Action) -> Result<Option<Action>> {
    let _ = action == Action::Tick;
    Ok(None)
  }

  fn init(&mut self, _area: ratatui::layout::Rect) -> Result<()> {
    Ok(())
  }

  fn handle_events(&mut self, event: Option<crate::tui::Event>) -> Result<Option<Action>> {
    if let Some(crate::tui::Event::Key(key)) = event {
      self.handle_key_events(key)
    } else {
      Ok(None)
    }
  }

  fn handle_key_events(&mut self, _key: KeyEvent) -> Result<Option<Action>> {
    Ok(None)
  }

  fn draw(&mut self, f: &mut Frame<'_>, area: Rect) -> Result<()> {
    f.render_widget(Paragraph::new("hello world"), area);
    Ok(())
  }
}
