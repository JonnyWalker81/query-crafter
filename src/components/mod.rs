pub mod custom_vim_editor;
pub mod db;
pub mod fps;
pub mod home;
pub mod vim;

use color_eyre::eyre::Result;
use crossterm::event::KeyEvent;
use ratatui::{layout::Rect, Frame};
use serde::{Deserialize, Serialize};
use tokio::sync::mpsc::UnboundedSender;

use crate::{action::Action, config::Config, tui};

#[derive(Default, Debug, Copy, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum ComponentKind {
  #[default]
  Home,
  Query,
  Results,
}

pub trait Component {
  fn register_action_handler(&mut self, tx: UnboundedSender<Action>) -> Result<()>;
  fn register_config_handler(&mut self, config: Config) -> Result<()>;
  fn update(&mut self, action: Action) -> Result<Option<Action>>;
  fn handle_events(&mut self, event: Option<tui::Event>) -> Result<Option<Action>>;
  fn handle_key_events(&mut self, key: KeyEvent) -> Result<Option<Action>>;
  fn init(&mut self, area: Rect) -> Result<()>;
  fn draw(&mut self, f: &mut Frame<'_>, area: Rect) -> Result<()>;
}
