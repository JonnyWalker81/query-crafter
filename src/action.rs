use serde::{Deserialize, Serialize};
use strum::Display;

use crate::components::{db::DbTable, ComponentKind};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Display, Deserialize)]
pub enum Action {
  Tick,
  Render,
  Resize(u16, u16),
  Suspend,
  Resume,
  Quit,
  Refresh,
  Error(String),
  Help,
  TablesLoaded(Vec<DbTable>),
  TableMoveUp,
  TableMoveDown,
  RowMoveUp,
  RowMoveDown,
  ScrollTableLeft,
  ScrollTableRight,
  LoadSelectedTable,
  LoadTables(String),
  LoadTable(String),
  QueryResult(Vec<String>, Vec<Vec<String>>),
  FocusQuery,
  FocusResults,
  FocusHome,
  SelectComponent(ComponentKind),
  ExecuteQuery,
  HandleQuery(String),
  RowDetails,
  SwitchEditor,
  ClearQuery,
}
