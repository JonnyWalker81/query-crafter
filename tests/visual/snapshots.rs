use insta::{assert_snapshot, with_settings};
use crate::test_utils::{ComponentTestHarness, TestEnvironment, fixtures};
use query_crafter::components::{db::Db, Component, ComponentKind};
use query_crafter::action::Action;
use ratatui::backend::TestBackend;
use ratatui::Terminal;

/// Helper to render component and return string representation
async fn render_component(component: &mut Db) -> String {
    let backend = TestBackend::new(80, 24);
    let mut terminal = Terminal::new(backend).unwrap();
    
    terminal.draw(|f| {
        component.draw(f, f.area()).unwrap();
    }).unwrap();
    
    let buffer = terminal.backend().buffer();
    let mut output = String::new();
    for y in 0..buffer.area.height {
        for x in 0..buffer.area.width {
            let cell = &buffer[(x, y)];
            output.push_str(cell.symbol());
        }
        output.push('\n');
    }
    output.trim_end().to_string()
}

#[tokio::test]
async fn test_initial_layout() {
    let env = TestEnvironment::new().await.unwrap();
    let mut db = Db::new();
    
    // Initialize component
    let (tx, _rx) = tokio::sync::mpsc::unbounded_channel();
    db.register_action_handler(tx).unwrap();
    db.register_config_handler(env.config).unwrap();
    db.init(ratatui::layout::Rect::new(0, 0, 80, 24)).unwrap();
    
    let output = render_component(&mut db).await;
    
    with_settings!({
        description => "Initial empty layout with three panels"
    }, {
        assert_snapshot!(output);
    });
}

#[tokio::test]
async fn test_tables_loaded_view() {
    let env = TestEnvironment::new().await.unwrap();
    let mut db = Db::new();
    
    // Initialize
    let (tx, _rx) = tokio::sync::mpsc::unbounded_channel();
    db.register_action_handler(tx).unwrap();
    db.register_config_handler(env.config).unwrap();
    db.init(ratatui::layout::Rect::new(0, 0, 80, 24)).unwrap();
    
    // Load tables
    let tables = fixtures::sample_tables();
    db.update(Action::TablesLoaded(tables)).unwrap();
    
    let output = render_component(&mut db).await;
    
    with_settings!({
        description => "Tables panel populated with sample tables"
    }, {
        assert_snapshot!(output);
    });
}

#[tokio::test]
async fn test_query_results_view() {
    let env = TestEnvironment::new().await.unwrap();
    let mut db = Db::new();
    
    // Initialize
    let (tx, _rx) = tokio::sync::mpsc::unbounded_channel();
    db.register_action_handler(tx).unwrap();
    db.register_config_handler(env.config).unwrap();
    db.init(ratatui::layout::Rect::new(0, 0, 80, 24)).unwrap();
    
    // Load query results
    let (headers, results) = fixtures::sample_query_results();
    db.update(Action::QueryResult(headers, results)).unwrap();
    
    let output = render_component(&mut db).await;
    
    with_settings!({
        description => "Results panel showing query results in table format"
    }, {
        assert_snapshot!(output);
    });
}

#[tokio::test]
async fn test_help_popup() {
    let env = TestEnvironment::new().await.unwrap();
    let mut db = Db::new();
    
    // Initialize
    let (tx, _rx) = tokio::sync::mpsc::unbounded_channel();
    db.register_action_handler(tx).unwrap();
    db.register_config_handler(env.config).unwrap();
    db.init(ratatui::layout::Rect::new(0, 0, 80, 24)).unwrap();
    
    // Show help by sending '?' key
    use crossterm::event::{KeyCode, KeyEvent, KeyEventKind, KeyEventState, KeyModifiers};
    let key_event = KeyEvent {
        code: KeyCode::Char('?'),
        modifiers: KeyModifiers::empty(),
        kind: KeyEventKind::Press,
        state: KeyEventState::empty(),
    };
    db.handle_key_events(key_event).unwrap();
    
    let output = render_component(&mut db).await;
    
    with_settings!({
        description => "Help popup overlay showing keyboard shortcuts"
    }, {
        assert_snapshot!(output);
    });
}

#[tokio::test]
async fn test_error_message_display() {
    let env = TestEnvironment::new().await.unwrap();
    let mut db = Db::new();
    
    // Initialize
    let (tx, _rx) = tokio::sync::mpsc::unbounded_channel();
    db.register_action_handler(tx).unwrap();
    db.register_config_handler(env.config).unwrap();
    db.init(ratatui::layout::Rect::new(0, 0, 80, 24)).unwrap();
    
    // Show error
    db.update(Action::Error("Database connection failed: timeout".to_string())).unwrap();
    
    let output = render_component(&mut db).await;
    
    with_settings!({
        description => "Error message displayed to user"
    }, {
        assert_snapshot!(output);
    });
}

#[tokio::test]
async fn test_query_editor_with_content() {
    let env = TestEnvironment::new().await.unwrap();
    let mut db = Db::new();
    
    // Initialize
    let (tx, _rx) = tokio::sync::mpsc::unbounded_channel();
    db.register_action_handler(tx).unwrap();
    db.register_config_handler(env.config).unwrap();
    db.init(ratatui::layout::Rect::new(0, 0, 80, 24)).unwrap();
    
    // Focus query editor and add content
    db.update(Action::FocusQuery).unwrap();
    
    // Simulate typing a query by updating the editor directly
    // (In real app this would come through key events)
    let query = "SELECT u.id, u.name, u.email\nFROM users u\nWHERE u.active = true\nORDER BY u.created_at DESC";
    
    // Since we can't easily simulate typing, we'll use a workaround
    // In a real test, you'd send key events
    
    let output = render_component(&mut db).await;
    
    with_settings!({
        description => "Query editor showing multi-line SQL query"
    }, {
        assert_snapshot!(output);
    });
}

#[tokio::test]
async fn test_table_columns_popup() {
    let env = TestEnvironment::new().await.unwrap();
    let mut db = Db::new();
    
    // Initialize
    let (tx, _rx) = tokio::sync::mpsc::unbounded_channel();
    db.register_action_handler(tx).unwrap();
    db.register_config_handler(env.config).unwrap();
    db.init(ratatui::layout::Rect::new(0, 0, 80, 24)).unwrap();
    
    // Load tables and columns
    let tables = fixtures::sample_tables();
    db.update(Action::TablesLoaded(tables)).unwrap();
    
    let columns = fixtures::sample_columns();
    db.update(Action::TableColumnsLoaded("users".to_string(), columns)).unwrap();
    
    // Show columns popup
    db.update(Action::ViewTableColumns).unwrap();
    
    let output = render_component(&mut db).await;
    
    with_settings!({
        description => "Table columns popup showing column details"
    }, {
        assert_snapshot!(output);
    });
}

#[tokio::test]
async fn test_focused_component_highlighting() {
    let env = TestEnvironment::new().await.unwrap();
    let mut db = Db::new();
    
    // Initialize
    let (tx, _rx) = tokio::sync::mpsc::unbounded_channel();
    db.register_action_handler(tx).unwrap();
    db.register_config_handler(env.config).unwrap();
    db.init(ratatui::layout::Rect::new(0, 0, 80, 24)).unwrap();
    
    // Load some data
    let tables = fixtures::sample_tables();
    db.update(Action::TablesLoaded(tables)).unwrap();
    
    // Focus different components and capture snapshots
    let components = vec![
        (ComponentKind::Home, "home_focused"),
        (ComponentKind::Query, "query_focused"),
        (ComponentKind::Results, "results_focused"),
    ];
    
    for (component, name) in components {
        db.update(Action::SelectComponent(component)).unwrap();
        let output = render_component(&mut db).await;
        
        with_settings!({
            description => format!("Layout with {} component focused", name)
        }, {
            assert_snapshot!(name, output);
        });
    }
}

#[tokio::test]
async fn test_long_results_scrolling() {
    let env = TestEnvironment::new().await.unwrap();
    let mut db = Db::new();
    
    // Initialize
    let (tx, _rx) = tokio::sync::mpsc::unbounded_channel();
    db.register_action_handler(tx).unwrap();
    db.register_config_handler(env.config).unwrap();
    db.init(ratatui::layout::Rect::new(0, 0, 80, 24)).unwrap();
    
    // Load long results
    let (headers, results) = fixtures::long_query_results();
    db.update(Action::QueryResult(headers, results)).unwrap();
    
    // Focus results
    db.update(Action::FocusResults).unwrap();
    
    // Move down several times to show scrolling
    for _ in 0..10 {
        db.update(Action::RowMoveDown).unwrap();
    }
    
    let output = render_component(&mut db).await;
    
    with_settings!({
        description => "Results panel showing scrolled view of long results"
    }, {
        assert_snapshot!(output);
    });
}