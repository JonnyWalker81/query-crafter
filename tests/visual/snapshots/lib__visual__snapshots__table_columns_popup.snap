---
source: tests/visual/snapshots.rs
description: Table columns popup showing column details
expression: output
---
╭──────────────────────────────────────────────────────────────────────────────╮
│Query Crafter - [1] Tables [2] Query [3] Results                              │
╰──────────────────────────────────────────────────────────────────────────────╯
╭[1] Tables────╮Query [t] History [t]                                           
│users  ╭Table Columns - users (ESC to close)──────────────────────────╮        
│posts  │                      Loading columns...                      │───────╮
│comment│                                                              │       │
│       │                                                              │       │
│       │                                                              │       │
│       │                                                              │       │
│       │                                                              │       │
│       │                                                              │       │
│       │                                                              │       │
│       │                                                              │       │
│       │                                                              │       │
│       │                                                              │       │
│       │                                                              │       │
│       │                                                              │       │
│       │                                                              │       │
│       ╰──────────────────────────────────────────────────────────────╯       │
│              ││                                                              │
│              ││                                                              │
│              │╰──────────────────────────────────────────────────────────────╯
╰──────────────╯╭[3] Results───────────────────────────────────────────────────╮
